### PR TITLE
Add defer , recover to handle panics

### PR DESCRIPTION
**Pull request by pzou1974**
fix of https://github.com/keikoproj/addon-manager/issues/18

Make a simple change to avoid the crash right before release.

### CODE REVIEW

File: controllers/addon_controller.go [hunk 1]
```diff
@@ -123,20 +123,28 @@ func (r *AddonReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
 		return reconcile.Result{}, ignoreNotFound(err)
 	}
+	
+	return r.execAddon(ctx, req, log, instance)
+}
 
+func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request, log logr.Logger, instance *addonmgrv1alpha1.Addon) (reconcile.Result, error) {
+	defer func() {
+		if err := recover(); err != nil {
+			fmt.Println("panic occurred during execAdd :", err)
```
Comment (kevdowney) [inline controllers/addon_controller.go]: Can we use the logger here?
```
log := r.Log
log.Info()
```

File: controllers/addon_controller.go [hunk 2]
```diff
@@ -123,20 +123,28 @@ func (r *AddonReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		}
+	}()
 	// Process addon instance
 	ret, err := r.processAddon(ctx, req, log, instance)
 
-	// Always update status, cache
+	// Always update cache, status
```
Comment (kevdowney) [inline controllers/addon_controller.go]: This comment was updated previously to represent the order of the code, we update status then cache.

File: controllers/addon_controller.go [hunk 3]
```diff
@@ -123,20 +123,28 @@ func (r *AddonReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+	defer func() {
+		if err := recover(); err != nil {
+			log.Info("Error: Panic occurred during execAdd %s/%s due to %s", instance.Namespace, instance.Name, err)
+		}
+	}()
 	// Process addon instance
 	ret, err := r.processAddon(ctx, req, log, instance)
 
-	// Always update status, cache
+	// Always update cache, status
```
Comment (kevdowney) [inline controllers/addon_controller.go]:
```suggestion
	// Always update status, cache
```

File: controllers/addon_controller.go [hunk 4]
```diff
@@ -123,20 +123,28 @@ func (r *AddonReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
 	err = r.updateAddonStatus(ctx, instance)
 	if err != nil {
 		// Force retry when status fails to update
 		ret.Requeue = true
 		return ret, err
 	}
 	r.addAddonToCache(instance)
-
-	return ret, err
+	return ret, nil
```
Comment (kevdowney) [inline controllers/addon_controller.go]: I think we found a logic error from past... if there's no panic but a legit error here we wanted to return the error from `processAddon`, can we change this return to `procErr` and above code to `procErr := r.processAddon()`?

Comment (pzou1974) [reply controllers/addon_controller.go]: why err is nil?  I thought there will be err, right?

Comment (pzou1974) [reply controllers/addon_controller.go]: agree. it makes more clear.

File: controllers/addon_controller.go [hunk 5]
```diff
@@ -123,19 +123,27 @@ func (r *AddonReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+}
 
+func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request, log logr.Logger, instance *addonmgrv1alpha1.Addon) (reconcile.Result, error) {
+	defer func() {
+		if err := recover(); err != nil {
+			log.Info("Error: Panic occurred during execAdd %s/%s due to %s", instance.Namespace, instance.Name, err)
+		}
+	}()
 	// Process addon instance
 	ret, err := r.processAddon(ctx, req, log, instance)
```
Comment (kevdowney) [inline controllers/addon_controller.go]:
```suggestion
	ret, procErr := r.processAddon(ctx, req, log, instance)
```

File: controllers/addon_controller.go [hunk 6]
```diff
@@ -123,19 +123,27 @@ func (r *AddonReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+	// Always update cache, status
 	err = r.updateAddonStatus(ctx, instance)
 	if err != nil {
 		// Force retry when status fails to update
 		ret.Requeue = true
 		return ret, err
 	}
 	r.addAddonToCache(instance)
-
 	return ret, err
```
Comment (kevdowney) [inline controllers/addon_controller.go]:
```suggestion
	return ret, procErr
```

Comment (pzou1974) [reply controllers/addon_controller.go]: change made

File: controllers/addon_controller.go (hunk 5 above)
Comment (pzou1974) [reply controllers/addon_controller.go]: this was from the original code but ok I will make change

**Pull request merged**